tests/unit/check_cl_dem_uart.c

```diff
@@ -0,0 +1,149 @@
+    ck_assert_int_eq(rv, OSD_OK);
+    ck_assert_uint_eq(dem_uart_desc.diaddr, dem_uart_diaddr);
+}
+END_TEST
+
+START_TEST(test_get_desc_wrong_module)
+{
+    osd_result rv;
+
+    mock_hostmod_expect_mod_describe(dem_uart_diaddr, OSD_MODULE_VENDOR_OSD,
+    								 OSD_MODULE_TYPE_STD_STM, 0);
+
+    struct osd_dem_uart_desc dem_uart_desc;
+    rv = osd_cl_dem_uart_get_desc(mock_hostmod_get_ctx(), dem_uart_diaddr,
+    							  &dem_uart_desc);
+    ck_assert_int_eq(rv, OSD_ERROR_WRONG_MODULE);
+}
+END_TEST
+
+static void dem_uart_handler(void *arg, struct osd_dem_uart_event *event)
+{
+    ck_assert(arg);
+    ck_assert(event);
+
+    // in this test we pass the expected event as a callback argument
+    struct osd_dem_uart_event *exp_event = arg;
+
+    ck_assert_uint_eq(exp_event->character, event->character);
+
+    free(event);
+}
+
+START_TEST(test_receive_event)
+{
+    osd_result rv;
+
+    struct osd_dem_uart_event_handler ev_handler;
+    ev_handler.cb_fn = dem_uart_handler;
+
+    struct osd_packet *pkg;
+    osd_packet_new(&pkg, osd_packet_sizeconv_payload2data(1));
+    rv = osd_packet_set_header(pkg, 1, 2, OSD_PACKET_TYPE_EVENT, EV_LAST);
+    ck_assert_int_eq(rv, OSD_OK);
+    pkg->data.payload[0] = 0x42;
+
+    struct osd_dem_uart_event exp_event;
+    exp_event.character = 0x42;
+    ev_handler.cb_arg = (void *) &exp_event;
+
+    rv = osd_cl_dem_uart_receive_event((void *) &ev_handler, pkg);
+    ck_assert_int_eq(rv, OSD_OK);
+}
+END_TEST
+
+START_TEST(test_send_events)
+{
+    osd_result rv;
+
+    struct osd_dem_uart_event event;
+    event.character = 0x42;
+
+    struct osd_dem_uart_ctx dem_uart_ctx;
+    dem_uart_ctx.hostmod_ctx = mock_hostmod_get_ctx();
+    dem_uart_ctx.local_diaddr = MOCK_HOSTMOD_DIADDR;
+    dem_uart_ctx.remote_diaddr = dem_uart_diaddr;
+
+    struct osd_packet *exp_packet;
+    osd_packet_new(&exp_packet, osd_packet_sizeconv_payload2data(1));
+    osd_packet_set_header(exp_packet, dem_uart_diaddr, MOCK_HOSTMOD_DIADDR,
+                          OSD_PACKET_TYPE_EVENT, EV_LAST);
+    exp_packet->data.payload[0] = 0x42;
+
+    mock_hostmod_expect_event_send(exp_packet, OSD_OK);
+
+    rv = osd_cl_dem_uart_send_event(&dem_uart_ctx, &event);
+    ck_assert_int_eq(rv, OSD_OK);
+}
+END_TEST
+
+Suite *suite(void)
+{
+    Suite *s;
+    TCase *tc_util, *tc_event_receive, *tc_event_send;
+
+    s = suite_create(TEST_SUITE_NAME);
+
+    tc_util = tcase_create("Utility Functionality");
+    tcase_add_checked_fixture(tc_util, setup, teardown);
+    tcase_add_test(tc_util, test_get_desc);
+    tcase_add_test(tc_util, test_get_desc_wrong_module);
+    suite_add_tcase(s, tc_util);
+
+    tc_event_receive = tcase_create("Handling of incoming events");
+    tcase_add_checked_fixture(tc_event_receive, setup, teardown);
+    tcase_add_test(tc_event_receive, test_receive_event);
+    suite_add_tcase(s, tc_event_receive);
+
+    tc_event_send = tcase_create("Generation of outgoing events");
```

Comment:
You don't need to have two test cases for the core functionality, you can call both tests in one tcase.